networkit/cpp/io/NetworkitBinaryReader.cpp

```diff
@@ -0,0 +1,150 @@
+
+namespace NetworKit {
+
+size_t NetworkitBinaryReader::decode(const uint8_t* data, uint64_t& value) {
+	int n;
+	if(!data[0]) {
+		n = 8;
+		value = 0;
+	} else {
+		n = tlx::ffs(data[0]) -1;
+		value = data[0] >> (n+1);
+	}
+
+	for(int i = 0; i < n; i++) {
+		value |= data[i+1] << (8 - (n + 1) + i * 8);
+	}
+	return n+1;
+}
+
+int64_t NetworkitBinaryReader::decodeZigzag(uint64_t value) {
+	int64_t orig = (value >> 1) ^ (-(value & 1));
+	return orig;
+}
+
+Graph NetworkitBinaryReader::read(const std::string& path) {
+	Header header = {};
+
+	MemoryMappedFile mmfile(path);
+	auto it = mmfile.cbegin();
+	auto end = mmfile.cend();
+
+	auto readHeader = [&] () {
+		memcpy(&header.magic, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+		memcpy(&header.checksum, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+		memcpy(&header.features, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+		memcpy(&header.nodes, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+		memcpy(&header.chunks, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+		memcpy(&header.offsetBaseData, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+		memcpy(&header.offsetAdjLists, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+		memcpy(&header.offsetWeights, it, sizeof(uint64_t));
+		it += sizeof(uint64_t);
+	};
+
+	auto checkHeader = [&] () {
+		if(memcmp("nkbg000", header.magic, 8)) {
+			throw std::runtime_error("Reader expected another magic value");
+		} else {
+			directed = (header.features & DIR_MASK);
+			weighted = (header.features & WGHT_MASK) >> WGHT_SHIFT;
+		}
+	};
+
+	readHeader();
+	checkHeader();
+
+	nodes = header.nodes;
+	DEBUG("# nodes here = ", nodes);
+	chunks = header.chunks;
+	DEBUG("# chunks here = ", chunks);
+	Graph G(nodes, weighted, directed);
+	std::vector<std::vector<uint64_t>> adj(nodes);
+
+	// Read base data
+	std::vector<uint64_t> prefSum;
+	std::vector<uint64_t> firstVert;
+	const char* base_it = mmfile.cbegin() + header.offsetBaseData;
+	for(uint64_t i = 0; i < nodes; i++) {
+		uint64_t sum = 0;
+		memcpy(&sum, base_it, sizeof(uint64_t));
+		base_it += sizeof(uint64_t);
+		prefSum.push_back(sum);
+	}
+
+	for(uint64_t ch = 0; ch < chunks; ch++) {
+		uint64_t first = 0;
+		memcpy(&first, base_it, sizeof(uint64_t));
+		base_it += sizeof(uint64_t);
+		firstVert.push_back(first);
+	}
+
+	// Read adjacency lists
+	const char* adj_it = mmfile.cbegin() + header.offsetAdjLists;
+
+	auto constructGraph = [&] (uint64_t c) {
+		node vertex = firstVert[c];
+
+		uint64_t off = 0;
+		memcpy(&off, adj_it + c * sizeof(uint64_t), sizeof(uint64_t));
+		off += chunks * sizeof(uint64_t);
+
+		uint64_t n;
+		if(c != (chunks -1)) {
+			n = firstVert[c+1] - vertex;
+		} else {
+			n = nodes - firstVert[c];
+		}
+
+		for(uint64_t i = 0; i < n; i++) {
+			uint64_t nbrs;
+			uint64_t curr = vertex+i;
+			uint64_t deleted;
+			if(curr > 0) {
+				nbrs = prefSum[curr] - prefSum[curr-1];
+				deleted = (nbrs & DELETED_BIT);
```

Comment:
This will not work. If (exactly) one of the `prefSum` entries has the `DELETED_BIT` bit set, `nbrs` is not computed correctly. To correctly handle this, mask out the `DELETED_BIT` when constructing `prefSum`.